build.rs

```diff
@@ -0,0 +1,384 @@
+                            }
+
+                            Some(make_impl_method(&m))
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+
+                let i = ModItem::Impl {
+                    target: type_name,
+                    items,
+                };
+
+                Some(i)
+            }
+
+            _ => None,
+        })
+        .filter(should_keep_item)
+        .collect()
+}
+
+fn should_keep_item(item: &ModItem) -> bool {
+    match item {
+        ModItem::Type { .. } => true,
+        ModItem::Trait { .. } => true,
+        ModItem::Impl { items, .. } => !items.is_empty(),
+        ModItem::Mod(m) => !m.items.is_empty(),
+    }
+}
+
+fn make_type_name(ty: &Box<syn::Type>) -> String {
+    ty.to_token_stream().to_string()
+}
+
+fn make_trait_method(m: &TraitItemMethod) -> Method {
+    let output = m.sig.output.to_token_stream();
+    let inputs = m
+        .sig
+        .inputs
+        .iter()
+        .map(|i| match i {
+            FnArg::Receiver(r) => r.to_token_stream().to_string(),
+            FnArg::Typed(ty) => ty.ty.to_token_stream().to_string(),
+        })
+        .collect::<Vec<_>>()
+        .join(", ");
+
+    Method {
+        name: m.sig.ident.to_string(),
+        tokens: format!("fn {}({}) {};", m.sig.ident, inputs, output),
+    }
+}
+
+fn make_impl_method(m: &ImplItemMethod) -> Method {
+    let output = m.sig.output.to_token_stream();
+    let inputs = m
+        .sig
+        .inputs
+        .iter()
+        .map(|i| match i {
+            FnArg::Receiver(r) => r.to_token_stream().to_string(),
+            FnArg::Typed(ty) => ty.ty.to_token_stream().to_string(),
+        })
+        .collect::<Vec<_>>()
+        .join(", ");
+
+    Method {
+        name: m.sig.ident.to_string(),
+        tokens: format!("fn {}({}) {};", m.sig.ident, inputs, output),
+    }
+}
+
+#[derive(Debug)]
+struct Mod {
+    name: String,
+    items: Vec<ModItem>,
+}
+
+impl Mod {
+    pub fn generate_code(&self, depth: usize) -> Result<String, std::fmt::Error> {
+        let mut code = String::new();
+
+        write!(code, "{}", TAB.repeat(depth))?;
+        writeln!(code, "mod {} {{", self.name)?;
+
+        for item in &self.items {
+            write!(code, "{}", item.generate_code(depth + 1)?)?;
+        }
+
+        write!(code, "{}", TAB.repeat(depth))?;
+        writeln!(code, "}}")?;
+        writeln!(code, "")?;
+
+        Ok(code)
+    }
+
+    pub fn generate_extern_crate(&self, depth: usize) -> Result<String, std::fmt::Error> {
+        let mut code = String::new();
+
+        write!(code, "{}", TAB.repeat(depth))?;
+        writeln!(code, "extern crate {} {{", self.name)?;
+
+        for item in &self.items {
+            write!(code, "{}", item.generate_code(depth + 1)?)?;
+        }
+
+        write!(code, "{}", TAB.repeat(depth))?;
+        writeln!(code, "}}")?;
+        writeln!(code, "")?;
+
```

Comment:
I see a lot of uses of ```write!``` and ```writeln!```. There is nothing inherently wrong with this approach, but it can also be done using ```quote!``` and ```rustfmt```. Example
```Rust
quote! {
    extern crate #name {
        #generated_code
    }
}
```
Then in main:
```Rust
let code = quote! {
    library! {
        #(#extern_crates)*
    }
}.to_string()
...
// Use rustfmt to format code
...
write!(f, formated_code);
```
See example in [codegen.file](https://github.com/dtolnay/syn/blob/master/codegen/src/file.rs) and [codegen.visit](https://github.com/dtolnay/syn/blob/master/codegen/src/visit.rs#L228)